api/tests/views_tests.py

```diff
@@ -104,6 +104,149 @@ def test_post_domainaddress_no_subdomain_error(premium_user, prem_api_client) ->
+    """
+    premium_profile = premium_user.profile
+    expected_num_address_deleted = premium_profile.num_address_deleted
+    expected_sent_welcome_email = premium_profile.sent_welcome_email
+
+    response = prem_api_client.patch(
+        reverse(viewname="profiles-detail", args=[premium_user.id]),
+        data={
+            "num_address_deleted": 5,
+            "sent_welcome_email": True,
+        },
+        format="json",
+    )
+
+    ret_data = response.json()
+    premium_profile.refresh_from_db()
+
+    assert premium_profile.num_address_deleted == expected_num_address_deleted
+    assert premium_profile.sent_welcome_email == expected_sent_welcome_email
+    assert ret_data.get("sent_welcome_email", [""])[0] == "This field is read only"
+    assert ret_data.get("num_address_deleted", [""])[0] == "This field is read only"
+    assert response.status_code == 400
+
+
+def test_profile_non_read_only_fields_update_correctly(premium_user, prem_api_client):
+    """
+    A field that is not read only should update correctly on a patch request.
+
+    "Not read only" meaning that it was defined in the serializers fields, but not read_only_fields.
+    """
+    premium_profile = premium_user.profile
+    old_onboarding_state = premium_profile.onboarding_state
+    old_email_tracker_remove_value = premium_profile.remove_level_one_email_trackers
+
+    response = prem_api_client.patch(
+        reverse(viewname="profiles-detail", args=[premium_user.id]),
+        data={
+            "onboarding_state": 1,
+            "remove_level_one_email_trackers": True,
+        },
+        format="json",
+    )
+
+    ret_data = response.json()
+    premium_profile.refresh_from_db()
+
+    assert (
+        ret_data.get("remove_level_one_email_trackers", None)
+        != old_email_tracker_remove_value
+    )
+    assert (
+        premium_profile.remove_level_one_email_trackers
+        != old_email_tracker_remove_value
+    )
+    assert premium_profile.remove_level_one_email_trackers is True
+    assert ret_data.get("remove_level_one_email_trackers", None) is True
+    assert ret_data.get("onboarding_state", None) != old_onboarding_state
+    assert premium_profile.onboarding_state == 1
+    assert ret_data.get("onboarding_state", None) == 1
+    assert response.status_code == 200
+
+
+def test_profile_patch_with_model_and_serializer_fields(premium_user, prem_api_client):
+    premium_profile = premium_user.profile
+
+    response = prem_api_client.patch(
+        reverse(viewname="profiles-detail", args=[premium_user.id]),
+        data={"subdomain": "vanilla", "num_address_deleted": 5},
+        format="json",
+    )
+
+    premium_profile.refresh_from_db()
+
+    assert response.status_code == 400
+    assert premium_profile.subdomain != "vanilla"
+    assert premium_profile.num_address_deleted == 0
+
+
+def test_profile_patch_with_non_read_only_and_read_only_fields(
```

Comment:
Case where the body of the request has both a read only, and a non-read-only field. Expected to return an error.